---
title: fgbio tools
---

# fgbio tools

The following tools are available in fgbio version 1.2.0.
## Basecalling

Tools for manipulating basecalling data.

|Tool|Description|
|----|-----------|
|[ExtractBasecallingParamsForPicard](ExtractBasecallingParamsForPicard.md)|Extracts sample and library information from an sample sheet for a given lane|
|[ExtractIlluminaRunInfo](ExtractIlluminaRunInfo.md)|Extracts information about an Illumina sequencing run from the RunInfo|

## FASTA

Tools for manipulating FASTA files.

|Tool|Description|
|----|-----------|
|[CollectAlternateContigNames](CollectAlternateContigNames.md)|Collates the alternate contig names from an NCBI assembly report|
|[HardMaskFasta](HardMaskFasta.md)|Converts soft-masked sequence to hard-masked in a FASTA file|
|[UpdateFastaContigNames](UpdateFastaContigNames.md)|Updates the sequence names in a FASTA|
|[UpdateIntervalListContigNames](UpdateIntervalListContigNames.md)|Updates the sequence names in an Interval List file|

## FASTQ

Tools for manipulating FASTQ files.

|Tool|Description|
|----|-----------|
|[DemuxFastqs](DemuxFastqs.md)|Performs sample demultiplexing on FASTQs|
|[FastqToBam](FastqToBam.md)|Generates an unmapped BAM (or SAM or CRAM) file from fastq files|
|[SortFastq](SortFastq.md)|Sorts a FASTQ file|
|[TrimFastq](TrimFastq.md)|Trims reads in one or more line-matched fastq files to a specific read length|

## RNA-Seq

Tools for RNA-Seq data

|Tool|Description|
|----|-----------|
|[CollectErccMetrics](CollectErccMetrics.md)|Collects metrics for ERCC spike-ins for RNA-Seq experiments|
|[EstimateRnaSeqInsertSize](EstimateRnaSeqInsertSize.md)|Computes the insert size for RNA-Seq experiments|

## SAM/BAM

Tools for manipulating SAM, BAM, or related data.

|Tool|Description|
|----|-----------|
|[AnnotateBamWithUmis](AnnotateBamWithUmis.md)|Annotates existing BAM files with UMIs (Unique Molecular Indices, aka Molecular IDs, Molecular barcodes) from a separate FASTQ file|
|[AutoGenerateReadGroupsByName](AutoGenerateReadGroupsByName.md)|Adds read groups to a BAM file for a single sample by parsing the read names|
|[ClipBam](ClipBam.md)|Clips reads from the same template|
|[ErrorRateByReadPosition](ErrorRateByReadPosition.md)|Calculates the error rate by read position on coordinate sorted mapped BAMs|
|[EstimatePoolingFractions](EstimatePoolingFractions.md)|Examines sequence data generated from a pooled sample and estimates the fraction of sequence data coming from each constituent sample|
|[ExtractUmisFromBam](ExtractUmisFromBam.md)|Extracts unique molecular indexes from reads in a BAM file into tags|
|[FilterBam](FilterBam.md)|Filters reads out of a BAM file|
|[FindSwitchbackReads](FindSwitchbackReads.md)|Finds reads where a template switch occurred during library construction|
|[FindTechnicalReads](FindTechnicalReads.md)|Find reads that are from technical or synthetic sequences in a BAM file|
|[RandomizeBam](RandomizeBam.md)|Randomizes the order of reads in a SAM or BAM file|
|[RemoveSamTags](RemoveSamTags.md)|Removes SAM tags from a SAM or BAM file|
|[SetMateInformation](SetMateInformation.md)|Adds and/or fixes mate information on paired-end reads|
|[SortBam](SortBam.md)|Sorts a SAM or BAM file|
|[SplitBam](SplitBam.md)|Splits a BAM into multiple BAMs, one per-read group (or library)|
|[TrimPrimers](TrimPrimers.md)|Trims primers from reads post-alignment|
|[UpdateReadGroups](UpdateReadGroups.md)|Updates one or more read groups and their identifiers|

## Unique Molecular Identifiers (UMIs)

Tools for manipulating UMIs & reads tagged with UMIs

|Tool|Description|
|----|-----------|
|[CallDuplexConsensusReads](CallDuplexConsensusReads.md)|Calls duplex consensus sequences from reads generated from the same double-stranded source molecule|
|[CallMolecularConsensusReads](CallMolecularConsensusReads.md)|Calls consensus sequences from reads with the same unique molecular tag|
|[CollectDuplexSeqMetrics](CollectDuplexSeqMetrics.md)|Collects a suite of metrics to QC duplex sequencing data|
|[CorrectUmis](CorrectUmis.md)|Corrects UMIs stored in BAM files when a set of fixed UMIs is in use|
|[FilterConsensusReads](FilterConsensusReads.md)|Filters consensus reads generated by CallMolecularConsensusReads or CallDuplexConsensusReads|
|[GroupReadsByUmi](GroupReadsByUmi.md)|Groups reads together that appear to have come from the same original molecule|
|[ReviewConsensusVariants](ReviewConsensusVariants.md)|Extracts data to make reviewing of variant calls from consensus reads easier|

## Utilities

Various utility programs.

|Tool|Description|
|----|-----------|
|[PickIlluminaIndices](PickIlluminaIndices.md)|Picks a set of molecular indices that should work well together|
|[PickLongIndices](PickLongIndices.md)|Picks a set of molecular indices that have at least a given number of mismatches between them|
|[UpdateDelimitedFileContigNames](UpdateDelimitedFileContigNames.md)|Updates the contig names in columns of a delimited data file (e|
|[UpdateGffContigNames](UpdateGffContigNames.md)|Updates then contig names in a GFF|

## VCF/BCF

Tools for manipulating VCF, BCF, or related data.

|Tool|Description|
|----|-----------|
|[AssessPhasing](AssessPhasing.md)|Assess the accuracy of phasing for a set of variants|
|[FilterSomaticVcf](FilterSomaticVcf.md)|Applies one or more filters to a VCF of somatic variants|
|[HapCutToVcf](HapCutToVcf.md)|Converts the output of 'HAPCUT' ('HapCut1'/'HapCut2') to a VCF|
|[MakeMixtureVcf](MakeMixtureVcf.md)|Creates a VCF with one sample whose genotypes are a mixture of other samples'|
|[MakeTwoSampleMixtureVcf](MakeTwoSampleMixtureVcf.md)|Creates a simulated tumor or tumor/normal VCF by in-silico mixing genotypes from two samples|
|[UpdateVcfContigNames](UpdateVcfContigNames.md)|Updates then contig names in a VCF|


